plugins/TriggerPrimitiveMaker.cpp

```diff
@@ -28,34 +30,106 @@ namespace dunedaq::trigger {
   register_command("start", &TriggerPrimitiveMaker::do_start);
-  register_command("stop", &TriggerPrimitiveMaker::do_stop);
+  register_command("stop",  &TriggerPrimitiveMaker::do_stop);
   register_command("scrap", &TriggerPrimitiveMaker::do_scrap);
+  // clang-format on
 }
 
 void
 TriggerPrimitiveMaker::init(const nlohmann::json& obj)
 {
-  m_tpset_sink.reset(new appfwk::DAQSink<TPSet>(appfwk::queue_inst(obj, "tpset_sink")));
+  m_init_obj = obj;
 }
 
 void
 TriggerPrimitiveMaker::do_configure(const nlohmann::json& obj)
 {
   m_conf = obj.get<triggerprimitivemaker::ConfParams>();
 
-  std::ifstream file(m_conf.filename);
+  // For each of the streams that are specified in the config, we read
+  // the input file, and create an outgoing sink. We also keep track
+  // of the total timestamp range of all the streams, so we can keep
+  // the timestamps of the multiple streams in sync when replaying,
+  // even when they don't all start or end at the same time
+  
+  m_earliest_first_tpset_timestamp = std::numeric_limits<triggeralgs::timestamp_t>::max();
+  m_latest_last_tpset_timestamp = 0;
+
+  for (auto& stream : m_conf.tp_streams) {
+    TPStream this_stream;
+    this_stream.tpset_sink =
+      std::make_unique<appfwk::DAQSink<TPSet>>(appfwk::queue_inst(m_init_obj, stream.output_sink_name));
+
+    this_stream.tpsets = read_tpsets(stream.filename, stream.region_id, stream.element_id);
+
+    m_earliest_first_tpset_timestamp =
+      std::min(m_earliest_first_tpset_timestamp, this_stream.tpsets.front().start_time);
+
+    m_latest_last_tpset_timestamp = std::max(m_latest_last_tpset_timestamp, this_stream.tpsets.back().start_time);
+
+    m_tp_streams.push_back(std::move(this_stream));
+  }
+}
+
+void
+TriggerPrimitiveMaker::do_start(const nlohmann::json& /*args*/)
+{
+  TLOG(TLVL_ENTER_EXIT_METHODS) << get_name() << ": Entering do_start() method";
+  m_running_flag.store(true);
+
+  // We need the wall-clock time at which we'll send out the TPSet
+  // with the earliest timestamp, so we can keep all of the output
+  // streams in sync. We pick "now" plus a bit, to allow time for all
+  // of the threads to start up
+  auto earliest_timestamp_time = std::chrono::steady_clock::now() + std::chrono::milliseconds(10);
+
+  for (auto& stream : m_tp_streams) {
+    m_threads.emplace_back(&TriggerPrimitiveMaker::do_work,
+                           this,
+                           std::ref(m_running_flag),
+                           std::ref(stream.tpsets),
+                           std::ref(stream.tpset_sink),
+                           earliest_timestamp_time);
+  }
+  TLOG(TLVL_ENTER_EXIT_METHODS) << get_name() << ": Exiting do_start() method";
+}
+
+void
+TriggerPrimitiveMaker::do_stop(const nlohmann::json& /*args*/)
+{
+  TLOG(TLVL_ENTER_EXIT_METHODS) << get_name() << ": Entering do_stop() method";
+  m_running_flag.store(false);
+  for (auto& thr : m_threads) {
+    thr.join();
```

Comment:
When persisting threads in a class, I like to use `std::unique_ptr<std::thread>` so that you can check that the thread was created using `thr != nullptr`. Even if you don't do that, I'd make sure to call `thr.joinable()`

Reply:
Done